### DISCUSSION TITLE
Trying to create large amount of interfaces

### DISCUSSION THREAD

**Discussion by Mailstorm-ctrl** [netbox-community/netbox#13659] [Q&A] [answered]
I thought this would be an easy task but I'm really not sure what I'm doing wrong. I imported several hundred VMs into netbox. Now I need to make interfaces for those VMs. But I can't seem to get the api to work for actually creating interfaces for my VMs.

``` python
response = requests.get(headers=header, url=base_url+vms+options, verify=False)
json_dict = response.json()
```

This works as expected. I got all of the VMs that are loaded in. Great.

I loop through each VM since I need to assign an interface (All VMs have one interface. There's only a dozen or 2 that have more than 1)
``` python
interfaces = []
for vm in json_dict['results']:
    if vm['primary_ip'] is None:
        interfaces.append(json.dumps({
            "name": "veth0",
            "virtual_machine": vm['id']
        }))
```

This makes what I expect but when I got to try and post/put/patch all I get is "{'non_field_errors': ['Expected a list of items but got type "dict".']}" The below is my latest try. I've tried json.dumps(interfaces), just interfaces, but I can't seem to understand why I can't make the interfaces. I'm looking at the docs for creating multiple objects here: https://demo.netbox.dev/static/docs/rest-api/overview/#creating-multiple-objects

```python
response = requests.patch(headers=header,url=base_url+'/virtualization/interfaces',verify=False,data=str(interfaces))
print(response.json())
```

I usually use a library to interact with APIs so this is really the first time I'm interacting directly with requests. So this might be something super obvious. I also wanted to look at the docs but going to /api/docs just results in a page not found.

**Comment by candlerb** [comment, marked as answer]
Firstly, to create interfaces, you have to post to `/api/virtualization/interfaces/` (note the trailing slash). And you need to POST not PATCH.

Secondly, `data=str(interfaces)` won't work. That will give you a Python string representation - in this case of a list of strings (which is also wrong). Use `json=interfaces` instead, and get rid of the json.dumps.

Something like this (untested):

```
interfaces = []
for vm in json_dict['results']:
    if vm['primary_ip'] is None:
        interfaces.append({
            "name": "veth0",
            "virtual_machine": vm['id']
        })

response = requests.post(headers=header,url=base_url+'/virtualization/interfaces/',verify=False,json=interfaces)
print(response.json())
```

> I also wanted to look at the docs but going to /api/docs just results in a page not found.

It's `/api/schema/swagger-ui/`, as described in the [documentation](https://docs.netbox.dev/en/stable/integrations/rest-api/#interactive-documentation).

**Reply by Mailstorm-ctrl** [reply]
You are a G...thank you. I didn't know that json= was a thing instead of data=. But for the documentation thing that appears to be a mismatch between my autofill and what's actually available. My autofill will put me at https://demo.netbox.dev/static/docs/rest-api/overview/ which appears to be outdated now.

**Reply by candlerb** [reply]
You didn't say what version of Netbox you're using. The URL I gave was for [v3.5.0](https://github.com/netbox-community/netbox/releases/v3.5.0) and later.

**Reply by Mailstorm-ctrl** [reply]
I'm using 3.6. But I realized that when I search netbox docs, my autofill somehow took me to the link I provided which is no longer correct. I'm not entirely sure what happened but either way I was looking at outdated information in regards to the doc link.
